usr/src/uts/common/io/usb/clients/ccid/ccid.c

```diff
@@ -0,0 +1,4426 @@
+	    CCID_REQUEST_GET_PARAMS, 0, 0, 0, &cc)) != 0) {
+		return (ret);
+	}
+
+	if ((ret = ccid_command_queue(ccid, cc)) != 0)
+		goto done;
+
+	ccid_command_poll(ccid, cc);
+
+	if (cc->cc_state != CCID_COMMAND_COMPLETE) {
+		ret = EIO;
+		goto done;
+	}
+
+	ccid_command_status_decode(cc, &crs, &cis, NULL);
+	if (crs != CCID_REPLY_STATUS_COMPLETE) {
+		if (cis == CCID_REPLY_ICC_MISSING) {
+			ret = ENXIO;
+		} else {
+			ret = EIO;
+		}
+		goto done;
+	}
+
+	/*
+	 * The protocol is in ch_param2 of the header.
+	 */
+	prot = ccid_command_resp_param2(cc);
+	mlen = ccid_command_resp_length(cc);
+	cpbuf = cc->cc_response->b_rptr + sizeof (ccid_header_t);
+
+	ret = 0;
+	switch (prot) {
+	case 0:
+		if (mlen < sizeof (ccid_params_t0_t)) {
+			ret = EOVERFLOW;
+			goto done;
+		}
+		*protp = ATR_P_T0;
+		bcopy(cpbuf, &paramsp->ccp_t0, sizeof (ccid_params_t0_t));
+		break;
+	case 1:
+		if (mlen < sizeof (ccid_params_t1_t)) {
+			ret = EOVERFLOW;
+			goto done;
+		}
+		*protp = ATR_P_T1;
+		bcopy(cpbuf, &paramsp->ccp_t1, sizeof (ccid_params_t1_t));
+		break;
+	default:
+		ret = ECHRNG;
+		break;
+	}
+
+done:
+	ccid_command_free(cc);
+	return (ret);
+}
+
+static void
+ccid_hw_error(ccid_t *ccid, ccid_intr_hwerr_t *hwerr)
+{
+	ccid_slot_t *slot;
+
+	/* Make sure the slot number is within range. */
+	if (hwerr->cih_slot >= ccid->ccid_nslots)
+		return;
+
+	slot = &ccid->ccid_slots[hwerr->cih_slot];
+
+	/* The only error condition defined by the spec is overcurrent. */
+	if (hwerr->cih_code != CCID_INTR_HWERR_OVERCURRENT)
+		return;
+
+	/*
+	 * The worker thread will take care of this situation.
+	 */
+	slot->cs_flags |= CCID_SLOT_F_INTR_OVERCURRENT;
+	ccid_worker_request(ccid);
+}
+
+static void
+ccid_intr_pipe_cb(usb_pipe_handle_t ph, usb_intr_req_t *uirp)
+{
+	mblk_t *mp;
+	size_t msglen, explen;
+	uint_t i;
+	boolean_t change;
+	ccid_intr_hwerr_t ccid_hwerr;
+	ccid_t *ccid = (ccid_t *)uirp->intr_client_private;
+
+	mp = uirp->intr_data;
+	if (mp == NULL)
+		goto done;
+
+	msglen = msgsize(mp);
+	if (msglen == 0)
+		goto done;
+
+	switch (mp->b_rptr[0]) {
+	case CCID_INTR_CODE_SLOT_CHANGE:
+		mutex_enter(&ccid->ccid_mutex);
+		ccid->ccid_stats.cst_intr_slot_change++;
+
+		explen = 1 + ((2 * ccid->ccid_nslots + (NBBY-1)) / NBBY);
+		if (msglen < explen) {
+			ccid->ccid_stats.cst_intr_inval++;
+			mutex_exit(&ccid->ccid_mutex);
+			goto done;
+		}
+
+		change = B_FALSE;
+		for (i = 0; i < ccid->ccid_nslots; i++) {
+			uint_t byte = (i * 2 / NBBY) + 1;
+			uint_t shift = i * 2 % NBBY;
+			uint_t present = 1 << shift;
+			uint_t delta = 2 << shift;
+
+			if (mp->b_rptr[byte] & delta) {
+				ccid_slot_t *slot = &ccid->ccid_slots[i];
+
+				slot->cs_flags &= ~CCID_SLOT_F_INTR_MASK;
+				slot->cs_flags |= CCID_SLOT_F_CHANGED;
+				if (mp->b_rptr[byte] & present) {
+					slot->cs_flags |= CCID_SLOT_F_INTR_ADD;
+				} else {
+					slot->cs_flags |= CCID_SLOT_F_INTR_GONE;
+				}
+				change = B_TRUE;
+			}
+		}
+
+		if (change) {
+			ccid_worker_request(ccid);
+		
```

Comment:
Does this refer to also CCID_SLOT_F_NEED_IO_TEARDOWN waiters? Either way a comment would be good, presuming those waiters *will* get woken up somehow.